File: cs/src/ConstraintSolver/CNSAT/Types/Clause.cs
using System;
using System.Collections.Generic;

namespace Alica.Reasoner.CNSAT
{
	public class Clause : IComparable<Clause>
	{
		protected List<Lit> literals = new List<Lit>();
		
		
		
		public void AddChecked(Lit l) {
			bool found = false;
			if (l.IsTemporary) {
				for(int i=0; i<this.Literals.Count; i++) {
					if (this.Literals[i].IsTemporary && l.Atom == this.Literals[i].Atom) {
						found = true;
						break;
					}					
				}
				if (!found) {
					Literals.Add(l);
				} //else { 	}
				return;
			}			
			for(int i=0; i<this.Literals.Count; i++) {
				if (l.Atom == this.Literals[i].Atom) {
					found = true;
				    if (l.Sign != this.Literals[i].Sign) {
						this.IsTautologic = true;
						this.Literals.Clear();
					}
					//else {}
					break;
				}					
			}
			if (!found) {
				Literals.Add(l);
			}
			return;

		}
		internal bool IsTautologic {
			get;
			private set;
		}
		internal bool IsFinished {
			get;
			set;
		}
		
		public Clause Clone() {
			Clause clone = new Clause();
			clone.Literals.AddRange(this.literals);
			clone.IsFinished = this.IsFinished;
			clone.IsTautologic = this.IsTautologic;
			return clone;
		}
		
		
		public Clause ()
		{
			this.Activity = 0;
			this.Satisfied = false;
			this.IsFinished = false;
			this.IsTautologic = false;
		
		}
		
		public void Add(Lit l) {
			literals.Add(l);
		}
		
		public int avgActivity() {
			int ret=0;
			foreach(Lit l in literals) {
				ret+=l.Var.Activity;
			}
			return ret/literals.Count;
		}
		
		public int CompareTo(Clause other)
    	{
			
        	int a = this.Activity;//avgActivity();
			int b = other.Activity; //avgActivity();
			if(a<b) return 1;
			else if(a>b) return -1;
			else return 0;
    	}
		
		public bool checkSatisfied() {
			foreach(Lit l in literals) {
				if(l.Var.Assignment==l.Sign) {
					//this.Satisfied = true;
					return true;
				}
			}
			//this.Satisfied = false;
			return false;
		}
		
		public List<Lit> Literals {
			get {return literals;}
			set {literals = value;}
		}
		public bool Satisfied {set; get;}
		public Watcher[] watcher = new Watcher[2];
		public Var LastModVar {set; get;}
		public int Activity {set; get;}
		
		public void Print() {
			foreach(Lit l in Literals) {
				if(l.Sign==Assignment.False) Console.Write("-");
				if(l.Sign==Assignment.True) Console.Write("+");
				if(l.Sign==Assignment.Unassigned) Console.Write("#");
				Console.Write(l.Var.Index);
				if((watcher[0] != null && watcher[0].Lit == l) || (watcher[1] != null && watcher[1].Lit == l)) Console.Write("w");
				Console.Write(" ");
			}
			Console.WriteLine();
		}
	}
}

File: java/src/main/java/Reasoner/CNSAT/Clause.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:46
//

package Reasoner.CNSAT;

import java.util.LinkedList;
import java.util.List;

public class Clause implements Comparable<Clause>
{
    protected List<Lit> literals = new LinkedList<>();
    public void addChecked(Lit l) throws Exception {
        boolean found = false;
        if (l.getIsTemporary())
        {
            for (int i = 0;i < this.getLiterals().size();i++)
            {
                if (this.getLiterals().get(i).getIsTemporary() && l.getAtom() == this.getLiterals().get(i).getAtom())
                {
                    found = true;
                    break;
                }
                 
            }
            if (!found)
            {
                getLiterals().add(l);
            }
             
            return ;
        }
         
        for (int i = 0;i < this.getLiterals().size();i++)
        {
            //else { 	}
            if (l.getAtom() == this.getLiterals().get(i).getAtom())
            {
                found = true;
                if (l.getSign() != this.getLiterals().get(i).getSign())
                {
                    this.setIsTautologic(true);
                    this.getLiterals().clear();
                }
                 
                break;
            }
             
        }
        //else {}
        if (!found)
        {
            getLiterals().add(l);
        }
         
        return;
    }

    private boolean IsTautologic;
    public boolean getIsTautologic() {
        return IsTautologic;
    }

    public void setIsTautologic(boolean value) {
        IsTautologic = value;
    }

    private boolean IsFinished;
    public boolean getIsFinished() {
        return IsFinished;
    }

    public void setIsFinished(boolean value) {
        IsFinished = value;
    }

    public Clause clone() {
        try
        {
            Clause clone = new Clause();
            clone.getLiterals().addAll(this.literals);
            clone.setIsFinished(this.getIsFinished());
            clone.setIsTautologic(this.getIsTautologic());
            return clone;
        }
        catch (RuntimeException exception)
        {
            throw exception;
        }
        catch (Exception exception)
        {
            throw new RuntimeException(exception);
        }
    
    }

    public Clause() throws Exception {
        this.setActivity(0);
        this.setSatisfied(false);
        this.setIsFinished(false);
        this.setIsTautologic(false);
    }

    public void add(Lit l) throws Exception {
        literals.add(l);
    }

    public int avgActivity() throws Exception {
        int ret = 0;
        for (Lit l : literals)
        {
            ret += l.getVar().getActivity();
        }
        return ret / literals.size();
    }

    public int compareTo(Clause other) {
        int a = this.getActivity();
        int b = other.getActivity();
        if (a < b)
            return 1;
        else if (a > b)
            return -1;
        else
            return 0;  
    }

    public boolean checkSatisfied() throws Exception {
        for (Lit l : literals)
        {
            if (l.getVar().getAssignment() == l.getSign())
            {
                return true;
            }
             
        }
        return false;
    }


    public List<Lit> getLiterals() throws Exception {
        return literals;
    }

    public void setLiterals(List<Lit> value) throws Exception {
        literals = value;
    }

    private boolean Satisfied;
    public void setSatisfied(boolean value) {
        Satisfied = value;
    }

    public boolean getSatisfied() {
        return Satisfied;
    }

    public Watcher[] watcher = new Watcher[2];
    private Var LastModVar;
    public void setLastModVar(Var value) {
        LastModVar = value;
    }

    public Var getLastModVar() {
        return LastModVar;
    }

    private int Activity;
    public void setActivity(int value) {
        Activity = value;
    }

    public int getActivity() {
        return Activity;
    }

    public void print() throws Exception {
        for (Lit l : getLiterals())
        {
            if (l.getSign() == Assignment.False)
                System.out.print("-");
             
            if (l.getSign() == Assignment.True)
                System.out.print("+");
             
            if (l.getSign() == Assignment.Unassigned)
                System.out.print("#");

            System.out.print(l.getVar().getIndex());
            if ((watcher[0] != null && watcher[0].getLit() == l) || (watcher[1] != null && watcher[1].getLit() == l))
                System.out.print("w");

            System.out.print(" ");
        }
        System.out.println();
    }

}


